kt/godot-runtime/src/main/kotlin/godot/core/Constructors.kt

```diff
@@ -1,26 +1,121 @@
+@file:Suppress("UNCHECKED_CAST", "unused")
+
 package godot.core
 
+import godot.util.VoidPtr
+import godot.util.threadLocal
+
 abstract class KtConstructor<T: KtObject>(
-    val parameterCount: Int
+    vararg argsTypes: Pair<VariantType, Boolean>
 ) {
-    abstract operator fun invoke(vararg args: Any): T
+    val parameterCount: Int = argsTypes.size
+    val parameterTypes: Array<VariantType> = argsTypes.map { it.first }.toTypedArray()
+    val parameterNullables: Array<Boolean> = argsTypes.map { it.second }.toTypedArray()
+
+    abstract operator fun invoke(): T
+
+    fun construct(rawPtr: VoidPtr, instanceId: Long) = KtObject.instantiateWith(rawPtr, instanceId) {
+        val argsSize = TransferContext.buffer.int
+        require(argsSize == parameterCount) {
+            "Expecting $parameterCount parameter(s) for constructor, but got $argsSize instead."
+        }
+        for (i in 0 until parameterCount) {
+            paramsArray[i] = TransferContext.readSingleArgument(parameterTypes[i], parameterNullables[i])
+        }
+        TransferContext.buffer.rewind()
+        val instance = invoke()
+        resetParamsArray()
+        instance
+    }
+
+    companion object {
+        val paramsArray by threadLocal { arrayOf<Any?>(null, null, null, null, null) }
```

Comment:
I would make use of the constant `CONSTRUCTOR_MAX_ARGS` to init the array so we don't need to modify that part if we change its value.